pypsa/statistics.py

```diff
@@ -1237,7 +1238,7 @@ def market_value(
             bus_carrier=bus_carrier,
             nice_names=nice_names,
         )
-        df = self.revenue(**kwargs) / self.supply(**kwargs)
+        df = self.revenue(**kwargs) / self.supply(**{**kwargs, "at_port": False})
```

Comment:
One idea would be to use `energy_balance` here to not get 0 values for `supply` of links because p0 is always negative. 
So maybe we could change this line to :
```python
update_kwargs = {"at_port": False} if bus_carrier is None else {}
df = self.revenue(**kwargs) / self.energy_balance(**{**kwargs, **update_kwargs}).abs()
```

Reply:
Then we get the market_value in units of bus0 by default, and only when selecting a bus_carrier do we get the market_value for this bus_carrier in units of the bus_carrier.

For example, for electrolysis, where bus0 is electricity and bus1 is hydrogen:

- Default (bus_carrier=None) market_value: Revenue considering the purchase price of electricity and the selling price of hydrogen (positive revenue), divided by the utilization of p0, i.e., electricity consumption. Hence, we get €/MWh_el.
- bus_carrier="AC" (electricity): Revenue only considering the purchase price of electricity (negative revenue), divided by electricity consumption. Hence, we again get €/MWh_el, but negative, because for electricity we usually need to pay, unless prices are negative.
- bus_carrier="H2": Revenue only considering hydrogen, which is sold (positive revenue), divided by hydrogen generation. Here we get €/MWh_H2, which is positive.